lib/engine-object-file/src/artifact.rs

```diff
@@ -0,0 +1,647 @@
+        let function_body_lengths = function_body_inputs
+            .keys()
+            .map(|_function_body| 0u64)
+            .collect::<PrimaryMap<LocalFunctionIndex, u64>>();
+
+        let metadata = ModuleMetadata {
+            compile_info,
+            prefix: engine_inner.get_prefix(&data),
+            data_initializers,
+            function_body_lengths,
+        };
+
+        // let wasm_info = generate_data_structures_for_c(&metadata);
+        // generate_c(wasm_info);
+        /*
+        In the C file we need:
+        - imports
+        - exports
+
+        to construct an api::Module which is a Store (can be passed in via argument) and an
+        Arc<dyn Artifact> which means this struct which includes:
+        - CompileModuleInfo
+        - Features
+        - ModuleInfo
+        - MemoryIndex -> MemoryStyle
+        - TableIndex -> TableStyle
+        - LocalFunctionIndex -> FunctionBodyPtr // finished functions
+        - FunctionIndex -> FunctionBodyPtr // finished dynamic function trampolines
+        - SignatureIndex -> VMSharedSignatureindextureIndex // signatures
+         */
+
+        let serialized_data = bincode::serialize(&metadata).map_err(to_compile_error)?;
+        let mut metadata_binary = vec![0; 10];
+        let mut writable = &mut metadata_binary[..];
+        leb128::write::unsigned(&mut writable, serialized_data.len() as u64)
+            .expect("Should write number");
+        metadata_binary.extend(serialized_data);
+        let metadata_length = metadata_binary.len();
+
+        let maybe_obj_bytes = compiler.experimental_object_file_compile_module(
+            &target,
+            &metadata.compile_info,
+            module_translation.as_ref().unwrap(),
+            &function_body_inputs,
+            &metadata,
+            &metadata_binary,
+        );
+
+        let obj_bytes = if let Some(obj_bytes) = maybe_obj_bytes {
+            obj_bytes?
+        } else {
+            let compilation = compiler.compile_module(
+                &target,
+                &metadata.compile_info,
+                module_translation.as_ref().unwrap(),
+                function_body_inputs,
+            )?;
+            let mut obj = get_object_for_target(&target_triple).map_err(to_compile_error)?;
+            emit_data(&mut obj, WASMER_METADATA_SYMBOL, &metadata_binary)
+                .map_err(to_compile_error)?;
+            emit_compilation(&mut obj, compilation, &metadata, &target_triple)
+                .map_err(to_compile_error)?;
+            obj.write().map_err(to_compile_error)?
+        };
+
+        //let host_target = Triple::host();
+        //let is_cross_compiling = target_triple != &host_target;
+
+        Self::from_parts_crosscompiled(&mut *engine_inner, metadata, obj_bytes, metadata_length)
+    }
+
+    /// Generate the header file that goes with the generated object file.
+    pub fn generate_header_file(&self) -> String {
```

Comment:
I believe we can put this as a utility function outside of the Artifact, this will help on having testing, for example. Or using that header-function in an isolated way.

A function that given a `ModuleInfo` returns the String.
I think we could generate everything needed from that `ModuleInfo` (rather than taking &self).

Reply:
Hmm, well we need all the symbol names to put into the header file which is currently part of the engine/artifact, maybe there's a way to make that work though

Reply:
Just added a comment in the PR that guides on how to solve that (hopefully).

https://github.com/wasmerio/wasmer/pull/1585#issuecomment-686587245